### ISSUE TITLE
Virtual Machine NFS writes do not trigger

### ISSUE THREAD

**Issue by chadfurman** [facebook/watchman#201]
Using Vagrant with NFS, writes on my debian 8 host do not trigger watchman re-builds in the Ubuntu guest.  Running watchman 3.9.0 because Ember CLI demands 3.x.x T_T

**Comment by wez** [comment]
I don't think this is something that will ever work in a scalable way with any watcher.

We allude to this here, in an admittedly understated way:

https://facebook.github.io/watchman/docs/install.html#system-requirements
> Watchman relies on the operating system facilities for file notification, which means that you will likely have very poor results using it on any kind of remote or distributed filesystem.

All of the kernel file system change notification systems work by piggy backing into the code paths in the kernel where the changes happen.  They distribute only to processes running on that kernel instance.

While some systems theoretically support plumbing in change notifications from remote mounts, there is no support for that kind of thing in the NFS protocol, even if it is just used "locally" between a host and a guest VM.

**Comment by chadfurman** [comment]
Maybe not scalable, but it would work by triggering fs.stat checks on each file  -- slow, but possibly better than restarting `ember server` and having it rebuild everything.  That's what `nodemon -L` (legacy mode) does.

https://github.com/remy/nodemon/issues/146

**Comment by chadfurman** [comment]
You're right that the inotify triggers will likely not ever work:  https://www.virtualbox.org/ticket/10660

**Comment by wez** [comment]
Our primary goal with watchman is to be able to answer questions about how the filesystem has changed with as close to `O(changes)` (or better!) characteristics as we can get, so a polling solution with `O(tree-size)` and at the mercy of the network isn't top of our list.

I do recognize the value for these use cases at small scale, so I'm going to interpret this issue as "add a polling watcher".

I have some diffs in progress that will allow selecting a specific watcher implementation.  Once that lands, it will be pretty trivial for someone to add a polling watcher and make the appropriate configuration changes to use it.

I want to be transparent about that not being top of our priority list though.  If someone wants to take a stab at this, I would happily help them work through the PR.

**Comment by chadfurman** [comment]
Sounds good.  If you think it's going to be trivial, post me some instructions here and I'll gauge the LOE given my skill-level.

**Comment by wez** [comment]
I'll update this issue with a suggested plan of attack once the changesets I mentioned above have landed!

**Comment by wez** [comment]
Sorry for not getting back to this sooner.
https://github.com/facebook/watchman/commit/718d41f7e044dbf59609f9edf5cc50e68c42132b added the configuration option that will help unblock this issue.

For anyone interested in picking this up, here's the suggested plan of attack:

1. Look at `watchman_ops` in `watchman.h` to get a feel for the interface you'll need to implement.  You'll need to create `watcher/poll.c` and hook it up in `Makefile.am`.  There are several "methods" that can be implemented for a watcher.  For the polling watcher, you can make the the `(start|stop)_watch_(file|dir)` methods NOPs because you're going to be recursively watching the root and don't need any special handling on a per-file or per-dir basis.
2. The polling watcher should spawn a thread that performs the polling.  There are some thread synchronization concerns here; you can draw on the implementation of `watcher/fsevents.c` for managing signalling and starting and stopping threads.
3. The polling thread is conceptually very simple: in a loop, sleep for some configurable poll interval, then recursively walk the directory tree.  For each filename that has an mtime newer than the last loop crawl time, add the filename to a pending list (use `w_pending_coll_add` for that).  After crawling, if the pending list you built is not empty, use `pthread_cond_signal` to wake up the main watchman IO thread.  Look at the `fse_callback` function for inspiration.
4. Your implementations of `wait_notify` and `consume` notify can be similar to those functions in `fsevents.c`, but use the pending list you built instead of an ad-hoc list of entries that `fsevents.c` uses; the `consume_notify` method will use `w_pending_coll_append` to transfer the contents of your pending list into the one that the main IO thread will consume.

**Comment by woopla** [comment]
Let me add my voice/vote to this one. Now that Git has a hook for Watchman (since 2.16.0), I'd be very interested in a polling option. Our developers check out their code on NFS (connecting to a VNC host from their Windows laptops - corporate policy and all that 😸  ), and any way to speed up status, etc. operations would be of great help.

**Comment by wez** [comment]
As I mentioned above this use case is not anywhere in our priorities for the foreseeable future.  We are definitely happy to accept a pull request that adds this, and can provide guidance on getting that polished if someone from the community wants to take a stab at implementing this.

**Comment by deterralba** [comment]
I have a somewhat similar issue: I use flow on a vagrant ubuntu VM (on a linux host) and watchman was not detecting my updates.
My search engine brought me here, so I write this message for the others like me: I found a solution with https://github.com/mhallin/vagrant-notify-forwarder.

**Comment by QasimK** [comment]
Thanks @deterralba. For anyone else looking into this, `vagrant-notify-forwarder` seems to have [reliability issues](https://github.com/mhallin/vagrant-notify-forwarder/issues/15), and I've encountered them myself. Another alternative could be [watchdog](https://github.com/gorakhargosh/watchdog), specifically the included `watchmedo` shell utility, but I have not tried that myself yet.

**Comment by muratgozel** [comment]
Both watchman and watchmedo doesn't see changes inside the shared folders with vm. Alternatively, we can use `vagrant -c "COMMAND TO EXECUTE ON VM"` to execute command when a certain file changed in the local machine.

**Comment by talionarwork** [comment]
I too faced this issue when tried to run react-native development inside Docker container. Watchman can see changes to files inside mounted volume made only inside container, but not on the host :( If I will run `touch FILENAME` watchman will catch up with changes but I't really annoying to run this every time on any change.

**Comment by AgentCosmic** [comment]
Hi guys, I wrote a tool specifically to solve this issue. It's designed to send file changes over HTTP and it's cross platform too. Have a look :) https://github.com/AgentCosmic/xnotify
Hope this can help you guys until a better method is implemented in watchman.

**Comment by polarathene** [comment]
I'm not sure if it meets the same needs someone might use Watchman for, but [`watchexec` has a polling option](https://github.com/watchexec/watchexec/blob/main/doc/watchexec.1.ronn#behaviour-options) for usage with remote filesystems like NFS. Might be a viable alternative.
